File: packages/plugin-sdk-python/src/bookclerk_plugin_sdk/sparse_workerd/smoke.py
"""Out-of-tree workerd plugin smoke: ensure → materialize → describe + health.

Spawns the pinned Cloudflare ``workerd`` with a materialized Cap'n Proto config
and POSTs ``describe`` (plus a ``POST /invoke`` ``health`` call for storefront /
remoteLibrary guests) to the HTTP bridge. Does not require the Rust
``bookclerk-workerd`` binary.
"""

from __future__ import annotations

import json
import os
import secrets
import socket
import subprocess
import time
import tomllib
import urllib.error
import urllib.request
from pathlib import Path
from typing import Any

from ..tools import validate_manifest
from .config import materialize_config
from .ensure import default_cache_dir, ensure_workerd


def _free_loopback_port() -> int:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.bind(("127.0.0.1", 0))
        return int(sock.getsockname()[1])


def _wait_for_health(base: str, token: str, timeout_s: float = 15.0) -> None:
    url = f"{base}/health"
    deadline = time.monotonic() + timeout_s
    while True:
        try:
            req = urllib.request.Request(
                url,
                headers={"Authorization": f"Bearer {token}"},
                method="GET",
            )
            with urllib.request.urlopen(req, timeout=1.0) as resp:  # noqa: S310 — loopback
                if 200 <= resp.status < 300:
                    return
        except (urllib.error.URLError, TimeoutError, OSError):
            pass
        if time.monotonic() > deadline:
            raise TimeoutError(f"timeout waiting for {url}")
        time.sleep(0.05)


def _invoke_health(base: str, iface: str, token: str) -> Any:
    """``POST /invoke`` health probe for ``ContentSource`` / ``RemoteLibrary``.

    Args:
        base: Bridge base URL.
        iface: Cap'n interface name (``ContentSource`` or ``RemoteLibrary``).
        token: Bridge bearer token.

    Returns:
        The ``HealthOk`` dict from the reply union.

    Raises:
        RuntimeError: On a non-200 transport answer or an ``err`` reply.
    """
    from .. import _wire

    if iface == "ContentSource":
        params = _wire._content_source_health_params_codec
        results = _wire._content_source_health_results_codec
    else:
        params = _wire._remote_library_health_params_codec
        results = _wire._remote_library_health_results_codec
    req = urllib.request.Request(
        f"{base}/invoke",
        data=_wire._encode_message(params, {}, _wire.NO_CAPS),
        headers={
            "content-type": "application/x-capnp",
            "Authorization": f"Bearer {token}",
            "x-bookclerk-interface": iface,
            "x-bookclerk-method": "health",
            "x-bookclerk-context": json.dumps({"invocation": {"id": "smoke"}}),
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=30.0) as resp:  # noqa: S310 — loopback
            body = resp.read()
    except urllib.error.HTTPError as err:
        text = err.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"bridge HTTP {err.code}: {text}") from err
    reply = _wire._decode_message(results, body)["result"]
    if reply.get("kind") == "err":
        failure = reply.get("value") or {}
        raise RuntimeError(
            f"{iface}.health failed: {failure.get('code', 'internal')}: "
            f"{failure.get('message', 'plugin error')}"
        )
    return reply.get("value")


def _post_json(url: str, body: dict[str, Any], token: str) -> Any:
    data = json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        headers={
            "content-type": "application/json",
            "Authorization": f"Bearer {token}",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=30.0) as resp:  # noqa: S310 — loopback
            text = resp.read().decode("utf-8")
            status = resp.status
    except urllib.error.HTTPError as err:
        text = err.read().decode("utf-8", errors="replace")
        status = err.code
        if status != 400:
            raise RuntimeError(f"bridge HTTP {status}: {text}") from err
    value = json.loads(text)
    if isinstance(value, dict) and value.get("error"):
        err = value["error"]
        raise RuntimeError(
            f"POST {url} failed: {err.get('code', 'internal')}: "
            f"{err.get('message', 'bridge error')}"
        )
    return value


def run_smoke(plugin_dir: Path) -> str:
    """Smoke a workerd plugin without the Rust ``bookclerk-workerd`` binary.

    Args:
        plugin_dir: Path to a ``runtime = "workerd"`` plugin root.

    Returns:
        Multi-line status including plugin id and JSON describe/health detail.

    Raises:
        FileNotFoundError: If ``plugin.toml`` is missing.
        ValueError: If the manifest is invalid or not a workerd plugin.
        TimeoutError: If the bridge health endpoint never becomes ready.
        RuntimeError: If the describe/health bridge call fails.

    Examples:
        >>> # print(run_smoke(Path("./my-workerd-plugin")))
    """
    root = plugin_dir.resolve()
    toml_path = root / "plugin.toml"
    if not toml_path.is_file():
        raise FileNotFoundError(f"missing plugin.toml in {root}")
    manifest = tomllib.loads(toml_path.read_text(encoding="utf-8"))
    validate_manifest(manifest)
    runtime = manifest.get("runtime") or "native"
    if runtime != "workerd":
        raise ValueError(f'smoke requires runtime = "workerd" (got {runtime!r})')

    workerd_bin = ensure_workerd(default_cache_dir())
    port = _free_loopback_port()
    bridge_token = secrets.token_hex(32)
    config_path, listen_addr = materialize_config(
        root,
        manifest,
        listen_port=port,
        bridge_token=bridge_token,
    )
    base = f"http://{listen_addr}"

    env = {**os.environ, "BOOKCLERK_PLUGIN_ROOT": str(root)}
    proc = subprocess.Popen(
        [str(workerd_bin), "serve", str(config_path)],
        cwd=str(root),
        stdin=subprocess.DEVNULL,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        env=env,
        text=True,
    )
    try:
        _wait_for_health(base, bridge_token)
        describe = _post_json(f"{base}/describe", {}, bridge_token)
        # `health` is a method of the storefront and remoteLibrary entrypoints;
        # the default entrypoint (event/job triggers) has no health probe.
        entrypoints = [str(e) for e in (manifest.get("entrypoints") or [])]
        if "storefront" in entrypoints:
            health_iface: str | None = "ContentSource"
        elif "remoteLibrary" in entrypoints:
            health_iface = "RemoteLibrary"
        else:
            health_iface = None
        health = (
            _invoke_health(base, health_iface, bridge_token) if health_iface else None
        )
        detail = {
            "plugin": manifest["id"],
            "listen": listen_addr,
            "describe": describe,
            "health": health,
        }
        return f"smoke ok {manifest['id']}\n{json.dumps(detail, indent=2)}"
    except Exception:
        try:
            out, err = proc.communicate(timeout=0.2)
            if out:
                print(out, flush=True)
            if err:
                print(err, flush=True)
        except Exception:  # noqa: BLE001
            pass
        raise
    finally:
        if proc.poll() is None:
            proc.terminate()
            try:
                proc.wait(timeout=3)
            except subprocess.TimeoutExpired:
                proc.kill()
                proc.wait(timeout=3)

File: packages/plugin-sdk/src/sparse-workerd/smoke.ts
/**
 * Out-of-tree workerd plugin smoke: ensure → materialize → describe + health.
 *
 * Spawns the pinned `workerd` binary against a materialised Cap'n Proto config
 * and exercises the bridge `/health`, `describe()`, and (for content-source /
 * integration kinds) the role `health` route.
 */

import fs from "node:fs";
import net from "node:net";
import path from "node:path";
import { randomBytes } from "node:crypto";
import { spawn, type ChildProcess } from "node:child_process";
import { parse as parseToml } from "smol-toml";
import { validateManifest, type Manifest } from "../tools/validate.js";
import { materializeConfig } from "./config.js";
import { defaultCacheDir, ensureWorkerd } from "./ensure.js";

async function freeLoopbackPort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const server = net.createServer();
    server.listen(0, "127.0.0.1", () => {
      const addr = server.address();
      if (!addr || typeof addr === "string") {
        server.close();
        reject(new Error("failed to allocate loopback port"));
        return;
      }
      const { port } = addr;
      server.close((err) => (err ? reject(err) : resolve(port)));
    });
    server.on("error", reject);
  });
}

async function waitForHealth(
  base: string,
  token: string,
  timeoutMs = 15_000,
): Promise<void> {
  const url = `${base}/health`;
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    try {
      const res = await fetch(url, {
        headers: { Authorization: `Bearer ${token}` },
      });
      if (res.ok) return;
    } catch {
      // retry
    }
    if (Date.now() > deadline) {
      throw new Error(`timeout waiting for ${url}`);
    }
    await new Promise((r) => setTimeout(r, 50));
  }
}

async function postJson(
  url: string,
  body: unknown,
  token: string,
): Promise<unknown> {
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(body ?? {}),
  });
  const text = await res.text();
  if (!res.ok && res.status !== 400) {
    throw new Error(`bridge HTTP ${res.status}: ${text}`);
  }
  const value = JSON.parse(text) as {
    error?: { code?: string; message?: string };
  };
  if (value.error) {
    throw new Error(
      `POST ${url} failed: ${value.error.code ?? "internal"}: ${value.error.message ?? "bridge error"}`,
    );
  }
  return value;
}

function loadManifest(pluginDir: string): Manifest {
  const tomlPath = path.join(pluginDir, "plugin.toml");
  if (!fs.existsSync(tomlPath)) {
    throw new Error(`missing plugin.toml in ${pluginDir}`);
  }
  const m = parseToml(fs.readFileSync(tomlPath, "utf8")) as Manifest;
  validateManifest(m);
  return m;
}

function killChild(child: ChildProcess): void {
  if (child.exitCode !== null || child.signalCode !== null) return;
  try {
    child.kill("SIGTERM");
  } catch {
    // ignore
  }
  setTimeout(() => {
    if (child.exitCode === null && child.signalCode === null) {
      try {
        child.kill("SIGKILL");
      } catch {
        // ignore
      }
    }
  }, 2000).unref?.();
}

/**
 * Smokes a workerd plugin directory without the Rust `bookclerk-workerd` binary.
 *
 * @param pluginDir - Plugin root with `runtime = "workerd"`.
 * @returns Human-readable success summary including describe/health JSON.
 * @throws {Error} When the runtime is not workerd, workerd fails to start, or
 *   bridge calls fail.
 *
 * @example
 * ```ts
 * console.log(await runSmoke("./my-plugin"));
 * ```
 */
export async function runSmoke(pluginDir: string): Promise<string> {
  const root = path.resolve(pluginDir);
  const manifest = loadManifest(root);
  const runtime = manifest.runtime ?? "native";
  if (runtime !== "workerd") {
    throw new Error(
      `smoke requires runtime = "workerd" (got ${JSON.stringify(runtime)})`,
    );
  }

  const workerdBin = await ensureWorkerd(defaultCacheDir());
  const port = await freeLoopbackPort();
  const bridgeToken = randomBytes(32).toString("hex");
  const generated = materializeConfig(root, manifest, {
    listenPort: port,
    notifyAddr: null,
    bridgeToken,
  });
  const base = `http://${generated.listenAddr}`;

  const child = spawn(workerdBin, ["serve", generated.configPath], {
    cwd: root,
    stdio: ["ignore", "pipe", "pipe"],
    env: { ...process.env, BOOKCLERK_PLUGIN_ROOT: root },
  });

  const logs: string[] = [];
  const onChunk = (buf: Buffer) => {
    const line = buf.toString("utf8").trimEnd();
    if (line) {
      logs.push(line);
      console.error(`workerd: ${line}`);
    }
  };
  child.stdout?.on("data", onChunk);
  child.stderr?.on("data", onChunk);

  try {
    await waitForHealth(base, bridgeToken);
    const describe = await postJson(`${base}/describe`, {}, bridgeToken);
    // Role `health` exists for content-source / integration kinds only.
    const healthPath =
      manifest.kind === "source"
        ? "/contentSource/health"
        : manifest.kind === "integration"
          ? "/integration/health"
          : null;
    const health = healthPath
      ? await postJson(`${base}${healthPath}`, {}, bridgeToken)
      : null;
    const detail = {
      plugin: manifest.id,
      listen: generated.listenAddr,
      describe,
      health,
    };
    return `smoke ok ${manifest.id}\n${JSON.stringify(detail, null, 2)}`;
  } catch (err) {
    if (logs.length) {
      console.error(`workerd logs (${logs.length} chunks) before failure`);
    }
    throw err;
  } finally {
    killChild(child);
    await new Promise<void>((resolve) => {
      if (child.exitCode !== null || child.signalCode !== null) {
        resolve();
        return;
      }
      child.once("exit", () => resolve());
      setTimeout(() => resolve(), 3000).unref?.();
    });
  }
}
